php/tests/ExampleCharacterTest.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace Emilybache\StrangecharacterTestdesignKata;
+
+use Emilybache\StrangecharacterTestdesignKata\Character;
+use Emilybache\StrangecharacterTestdesignKata\CharacterFinder;
+use PHPUnit\Framework\TestCase;
+
+class ExampleCharacterTest extends TestCase
+{
+    public function testFindCharacterByLastName(): void
+    {
+        // This test constructs all its own test data to make it clearer what it does
+        $karen = new Character("Karen", "Wheeler");
+        $mike = new Character("Mike", "Wheeler");
+
+        // this is example code showing kinds of assertion you could do on a Character
+        $nancy = new Character("Nancy", "Wheeler");
+        $this->assertSame("Nancy", $nancy->firstName);
+
+        $karen->addChild($nancy);
+        $karen->addChild($mike);
+        $finder = new CharacterFinder([$karen, $mike, $nancy]);
+
+        // This is the 'act' step
+        $charactersList = $finder->findFamilyByLastName("Wheeler");
+
+        // this is example code showing kinds of assertion you could do on a List of Characters
+        $this->assertNotEmpty($charactersList);
+        $this->assertCount(3, $charactersList);
+        $this->assertCharacterListContains(
+            [
+                new Character("Nancy", "Wheeler"),
+                new Character("Mike", "Wheeler"),
+                new Character("Karen", "Wheeler"),
+            ],
+            $charactersList
+        );
+        $this->assertCharacterListIsEqualTo(
+            [
+                new Character("Nancy", "Wheeler"),
+                new Character("Mike", "Wheeler"),
+                new Character("Karen", "Wheeler"),
+            ],
+            $charactersList
+        );
+    }
+
+    // Helper assertion functions to get around the missing Set class in PHP that compares characters by first name
+
+    /**
+     * @param Character[] $expectedCharacters
+     * @param Character[] $charactersList
+     */
+    public static function assertCharacterListContains(array $expectedCharacters, array $charactersList): void
```

Comment:
is it possible to use a test framework that has this kind of assertion built in?

Reply:
PHP does not allow for equality comparisons on a specific value (i.e. comparing on ´firstName´) like the other languages do. [RFCs to bring that feature to the language have been rejected](https://wiki.php.net/rfc/object-comparison). PHP either compares by-value using *all* properties (which breaks the tests when comparing two characters with the same name and different parent/child properties like in the test) or the object references (which also breaks the test which relies on a by-value comparison). PHP testing libraries don't offer custom equality checks, so I implemented my own domain-specific one. If you think that these assertions are detrimental to the spirit of the kata, here are some changes that I could do:

1. Use by-reference checks (nor creating new comparison objects)
2. Extract the existing assertions into a separate file, so all unit tests could use them. I could even write unit tests for the assertions themselves to make sure I've covered all the cases (incomplete set, same object, different objects with same firstname)
3. Same as 2, but use nested loops and a custom `equals` operator. I'd have to tweak the algorithm a bit which would either make it more complicated or the performance characteristics slightly worse (O(n^2) instead of O(2n)), but might stay closer to the other implementations.

What would you like?

Reply:
I've extracted and simplified the assertions, using PHP functions instead of custom implementations.

Reply:
Ok so seems like a restriction in the language itself that prevents the kind of generic assertions I find in other languages. What you've done seems a good compromise. One goal of the exercise is to write tests that fail with really clear and understandable error messages. This custom assertion code you've included will help with that.